mesonbuild/mintro.py

```diff
@@ -125,7 +131,160 @@ def list_target_files(target_name, coredata, builddata):
+        self.cross_file = cross_file
+        self.native_file = None
+        self.cmd_line_options = {}
+
+class BuildoptionsInterperter(astinterpreter.AstInterpreter):
+    # Interpreter to detect the options without a build directory
+    # Most of the code is stolen from interperter.Interpreter
+    def __init__(self, source_root, subdir, backend, cross_file=None, subproject='', subproject_dir='subprojects'):
+        super().__init__(source_root, subdir)
+
+        options = BuildoptionsOptionHelper(cross_file)
+        self.cross_file = cross_file
+        self.environment = environment.Environment(source_root, None, options)
+        self.subproject = subproject
+        self.subproject_dir = subproject_dir
+        self.coredata = self.environment.get_coredata()
+        self.option_file = os.path.join(self.source_root, self.subdir, 'meson_options.txt')
+        self.backend = backends.get_backend_from_name(backend, None)
+        self.default_options = {'backend': self.backend.name}
+
+        self.funcs.update({
+            'project': self.func_project,
+            'add_languages': self.func_add_languages
+        })
+
+    def detect_compilers(self, lang, need_cross_compiler):
+        comp, cross_comp = self.environment.detect_compilers(lang, need_cross_compiler)
+        if comp is None:
+            # Logging would destroy the JSON output
+            return None, None
+
+        self.coredata.compilers[lang] = comp
+        # Native compiler always exist so always add its options.
+        new_options = comp.get_options()
+        if cross_comp is not None:
+            self.coredata.cross_compilers[lang] = cross_comp
+            new_options.update(cross_comp.get_options())
+
+        optprefix = lang + '_'
+        for k, o in new_options.items():
+            if not k.startswith(optprefix):
+                raise RuntimeError('Internal error, %s has incorrect prefix.' % k)
+            self.coredata.compiler_options.setdefault(k, o)
+
+        return comp, cross_comp
+
+    def flatten_args(self, args):
+        # Resolve mparser.ArrayNode if needed
+        flattend_args = []
+        for i in args:
+            if isinstance(i, mparser.ArrayNode):
+                flattend_args += [x.value for x in i.args.arguments]
+            elif isinstance(i, str):
+                flattend_args += [i]
+            else:
+                pass
+        return flattend_args
+
+    def add_languages(self, args):
+        need_cross_compiler = self.environment.is_cross_build() and self.environment.cross_info.need_cross_compiler()
+        for lang in sorted(args, key=compilers.sort_clink):
+            lang = lang.lower()
+            if lang in self.coredata.compilers:
+                comp = self.coredata.compilers[lang]
+            else:
+                (comp, _) = self.detect_compilers(lang, need_cross_compiler)
+            if comp is None:
+                return False
+            self.add_base_options(comp)
+        return True
```

Comment:
The return value is never used. Also we can move `self.add_base_options` into the branch for newly added compilers and delete the code for `lang in self.coredata.compilers`

Reply:
Done with the last two commits.